src/main/scala/io/iohk/ethereum/vm/ProgramState.scala

```diff
@@ -126,6 +133,12 @@ case class ProgramState[W <: WorldStateProxy[W, S], S <: Storage[S]](
   def revert(data: ByteString): ProgramState[W, S] =
     copy(error = Some(RevertOccurs), returnData = data, halted = true)
 
+  def addAccessedAddress(addr: Address): ProgramState[W, S] =
```

Comment:
Note: There seems to be two conventions in the same class here for naming (`doSomething`: e.g. `refundGas` vs `withSomething`: e.g. `withAddressToDelete` ). 

I personally slightly prefer `withAccessedAddress` to be clear that the structure is immutable. But I would agree that it also adds another ambiguity about whether the new instance accessed addresses are replaced by the passed value or if the value is appended, so I'm not fixed on that one.

Reply:
I understand `with*` functions as replacing the old value with the new one in the newly created copy. `add*` functions don't replace, but add a new item to the already existing collection. 

To tell whether the function modifies the structure on which it is called it should be enough to check it's signature. A function which does the modification in place probably doesn't need to return anything (maybe a Boolean stating whether the update was successful?). A pure function without side effects is going to return the same type as the object on which it is called (as it is in this case).